js/web/lib/wasm/session-handler.ts

```diff
@@ -119,128 +119,165 @@ const getLoggingLevel = (loggingLevel: 'verbose'|'info'|'warning'|'error'|'fatal
+  Object.entries(options).forEach(([key, value]) => {
+    const name = (prefix) ? prefix + '.' + key : key;
+    if (typeof value === 'object') {
+      iterateExtraOptions(value as Record<string, unknown>, name, handler);
+    } else if (['string', 'number', 'boolean'].indexOf(typeof value) !== -1) {
+      handler(name, (value as string | number | boolean).toString());
     } else {
-      errorCode = wasm._OrtDisableCpuMemArena(sessionOptionsHandle);
-    }
-    if (errorCode !== 0) {
-      throw new Error(`Can't set a CPU memory arena as a session option. error code = ${errorCode}`);
+      throw new Error(`Can't handle extra config type: ${typeof value}`);
     }
+  });
+};
+
+const setSessionOptions = (options?: InferenceSession.SessionOptions): [number, number[]] => {
+  const wasm = getInstance();
+  const allocs: number[] = [];
+
+  const sessionOptions: InferenceSession.SessionOptions = (options !== undefined) ? options : {};
+
+  if (options === undefined || options.graphOptimizationLevel === undefined) {
+    sessionOptions.graphOptimizationLevel = 'all';
   }
+  const graphOptimizationLevel = getGraphOptimzationLevel(sessionOptions.graphOptimizationLevel!);
 
-  if (options.enableMemPattern !== undefined) {
-    if (options.enableMemPattern) {
-      errorCode = wasm._OrtEnableMemPattern(sessionOptionsHandle);
-    } else {
-      errorCode = wasm._OrtDisableMemPattern(sessionOptionsHandle);
-    }
-    if (errorCode !== 0) {
-      throw new Error(`Can't set a memory pattern as a session option. error code = ${errorCode}`);
-    }
+  if (options === undefined || options.enableCpuMemArena === undefined) {
+    sessionOptions.enableCpuMemArena = true;
   }
 
-  if (options.executionMode !== undefined) {
-    switch (options.executionMode) {
-      case 'sequential':
-        errorCode = wasm._OrtSetSessionExecutionMode(sessionOptionsHandle, 0);
-        break;
-      case 'parallel':
-        errorCode = wasm._OrtSetSessionExecutionMode(sessionOptionsHandle, 1);
-        break;
-      default:
-        throw new Error(`unsupported execution mode: ${options.executionMode}`);
-    }
-    if (errorCode !== 0) {
-      throw new Error(`Can't set an execution mode as a session option. error code = ${errorCode}`);
-    }
+  if (options === undefined || options.enableMemPattern === undefined) {
+    sessionOptions.enableMemPattern = true;
+  }
+
+  if (options === undefined || options.executionMode === undefined) {
+    sessionOptions.executionMode = 'sequential';
   }
+  const executionMode = getExecutionMode(sessionOptions.executionMode!);
 
-  if (options.logId !== undefined) {
+  let logIdDataOffset = 0;
+  if (options !== undefined && options.logId !== undefined) {
     const logIdDataLength = wasm.lengthBytesUTF8(options.logId) + 1;
-    const logIdDataOffset = wasm._malloc(logIdDataLength);
+    logIdDataOffset = wasm._malloc(logIdDataLength);
     wasm.stringToUTF8(options.logId, logIdDataOffset, logIdDataLength);
```

Comment:
we can use a helper function for allocation buffer for utf-8 encoded string (looks like it's used multiple times)